main/commons/src/main/java/org/cryptomator/common/mountpoint/MountPointChooser.java

```diff
@@ -0,0 +1,170 @@
+	 * the Volume and check for any obstacles that could hinder the Mount-Operation.
+	 * The Mountpoint is deemed "prepared" if it can be used to mount a Volume
+	 * without any further Filesystemactions or Userinteraction. If this is not possible,
+	 * this method should fail. In other words: This method should not return without
+	 * either failing or finalizing the preparation of the Mountpoint.
+	 * Generally speaking exceptions should be wrapped as
+	 * {@link InvalidMountPointException} to allow efficient handling by the caller.
+	 *
+	 * <p>Often the preparation of a Mountpoint involves creating files or others
+	 * actions that require cleaning up after the Volume is unmounted.
+	 * In this case developers should override the {@link #cleanup(Path)}-Method
+	 * and return {@code true} to the Volume to indicate that the
+	 * {@code #cleanup}-Method of this MPC should be called after unmount.
+	 *
+	 * <p><b>Please note:</b> If this method fails the entire
+	 * Mountpoint-Choosing-Operation is aborted without calling {@link #cleanup(Path)}
+	 * or any other MPCs. Therefore this method should do all necessary cleanup
+	 * before throwing the exception.
+	 *
+	 * @param mountPoint the Mountpoint chosen by {@link #chooseMountPoint()}
+	 * @return a boolean flag; true if cleanup is needed, false otherwise
+	 * @throws InvalidMountPointException if the preparation fails
+	 * @see #chooseMountPoint()
+	 * @see #cleanup(Path)
+	 */
+	default boolean prepare(Path mountPoint) throws InvalidMountPointException {
+		return false; //NO-OP
+	}
+
+	/**
+	 * Called by a {@link Volume} to do any cleanup needed after unmount.
+	 *
+	 * <p>This method is only called if the {@link #prepare(Path)}-Method of the same
+	 * MountPointChooser returned {@code true}. Typically developers want to
+	 * delete any files created prior to mount or do similar tasks.<br>
+	 * Exceptions in this method should be handled gracefully.
+	 *
+	 * @param mountPoint the Mountpoint that was prepared by {@link #prepare(Path)}
+	 * @see #prepare(Path)
+	 */
+	default void cleanup(Path mountPoint) {
+		//NO-OP
+	}
+
+	/**
+	 * The phases of the existing {@link MountPointChooser MountPointChoosers.}
+	 * <p>The {@code Phases} of the MPCs are attached to them in the
+	 * {@link MountPointChooserModule} by annotating them with the
+	 * {@link PhaseKey @PhaseKey-Annotation.}
+	 * <p>Each MPC must have a <b>unique</b> Phase that allows the Module to sort
+	 * the MPCs according to their phases' {@link Phase#getTiming()} timing.}
+	 * The timing must be defined by the developer to reflect a useful execution order.
+	 *
+	 * <p><b>Phase-Uniqueness:</b> Phases must be unique, meaning that they must not be used
+	 * to annotate more than <i>one</i> Providermethod. Define a new phase for each additional
+	 * MPC that is added to the Module. Timings can be reused; the order of Phases with equal
+	 * timings among themselves is undefined.
```

Comment:
If this is mandatory and you don't want developers to forget assigning a Phase, better make it a property of the MPC itself instead of an "externally" assigned value.

Btw: Make MPC extend `Comparable<MPC>` will get you exactly the behaviour described here, if you put them in a sorted collection.

Reply:
Resolved by 5dac594cc922d701f0147da5672335813edd6324